src/adapter/tests.rs

```diff
@@ -400,6 +400,116 @@ fn rustdoc_finds_statics() {
     }
 }
 
+#[test]
+fn rustdoc_modules() {
+    let path = "./localdata/test_data/modules/rustdoc.json";
+
+    let content = std::fs::read_to_string(path)
+        .with_context(|| format!("Could not load {path} file, did you forget to run ./scripts/regenerate_test_rustdocs.sh ?"))
+        .expect("failed to load rustdoc");
+
+    let crate_ = serde_json::from_str(&content).expect("failed to parse rustdoc");
+    let indexed_crate = IndexedCrate::new(&crate_);
+    let adapter = Arc::new(RustdocAdapter::new(&indexed_crate, None));
+
+    let mod_query = r#"
+{
+   Crate {
+       item {
+           ... on Module {
+               module: name @output
+               is_crate @output
+               is_stripped @output
+               item {
+                   member: name @output
+               }
+           }
+       }
+   }
+}
+"#;
+
+    let variables: BTreeMap<&str, &str> = BTreeMap::default();
+
+    let schema =
+        Schema::parse(include_str!("../rustdoc_schema.graphql")).expect("schema failed to parse");
+
+    #[derive(Debug, PartialOrd, Ord, PartialEq, Eq, serde::Deserialize)]
+    struct Output {
+        member: Option<String>,
+        module: String,
+        is_crate: bool,
+        is_stripped: bool,
+    }
+
+    let mut results: Vec<Output> =
+        trustfall::execute_query(&schema, adapter.clone(), mod_query, variables.clone())
+            .expect("failed to run query")
+            .map(|row| row.try_into_struct().expect("shape mismatch"))
+            .collect();
+    results.sort_unstable();
+
+    similar_asserts::assert_eq!(
+        vec![
+            Output {
+                member: None,
```

Comment:
Consider outputting `__typename` on the member items as well, to make it easier to verify that `None` member names aren't bugs and are instead item types that genuinely don't have their own names (like `impl` for example).

Reply:
I've tried to do this, but my trial-and-error didn't find a way to make this work well with the @fold I added. I would have liked to output `(member name, member type)` tuples, but I didn't know enough graphql.

Reply:
Unfortunately we don't have a way for that yet, it's on my todo list but I haven't gotten to it.

`@fold` will do produce parallel lists of each property separately, so their types in `Output` would be `Vec<String>` for each. Trustfall also guarantees that the lists have equal length and are parallel, in the sense that the i-th element in each list corresponds to the same vertex each time.

Reply:
Okay, I tried it that way, but I got: 

```
thread 'adapter::tests::rustdoc_modules' panicked at 'internal error: entered unreachable code: unexpected item.inner for item: Item { id: Id("0:9-0:11:1801"), crate_id: 0, name: None, span: Some(Span { filename: "src/outer.rs", begin: (4, 0), end: (4, 18) }), visibility: Public, docs: None, links: {}, attrs: [], deprecation: None, inner: Import(Import { source: "inner::T4", name: "T4", id: Some(Id("0:11:1801")), glob: false }) }', src/adapter/vertex.rs:62:22
```

As far as I can tell, I think this means that rustdoc_types::Import doesn't have a `__typename`, and so we can't examine it.  That's probably a bug, but not one I think I can fix in this branch.

Reply:
Ah that actually caught a different bug, see above.